crates/nri-resctrl-plugin/src/lib.rs

```diff
@@ -155,6 +232,111 @@ impl<P: FsProvider> ResctrlPlugin<P> {
             warn!("resctrl-plugin: failed to send event: {}", e);
         }
     }
+
+    // Create or fetch pod state and ensure group exists
+    fn handle_new_pod(&self, pod: &nri::api::PodSandbox) -> Option<String> {
+        let pod_uid = &pod.uid;
+        let mut st = self.state.lock().unwrap();
+        let ps = st.pods.entry(pod_uid.clone()).or_insert(PodState {
+            group_path: None,
+            total_containers: 0,
+            reconciled_containers: 0,
+        });
+        if ps.group_path.is_none() {
+            match self.resctrl.create_group(pod_uid) {
+                Ok(p) => {
+                    ps.group_path = Some(p.clone());
+                    Some(p)
+                }
+                Err(e) => {
+                    warn!("resctrl-plugin: failed to create group for pod {}: {}", pod_uid, e);
+                    None
+                }
+            }
+        } else {
+            ps.group_path.clone()
+        }
+    }
+
+    fn handle_new_container(&self, pod: &nri::api::PodSandbox, container: &nri::api::Container) {
+        let pod_uid = pod.uid.clone();
+        let group_path = match self.handle_new_pod(pod) {
+            Some(g) => g,
+            None => {
+                // Emit failure for pod without group
+                let ev = PodResctrlEvent::Added(PodResctrlAdded {
+                    pod_uid: pod_uid.clone(),
+                    group_path: None,
+                    state: AssignmentState::Failure,
+                    total_containers: 0,
+                    reconciled_containers: 0,
+                });
+                self.emit_event(ev);
+                return;
+            }
+        };
+
+        // Enumerate container PIDs and reconcile just this container into the pod group
+        let passes = self.cfg.max_reconcile_passes;
+        let pids = self.pid_source.pids_for_container(container);
+        let res = self
+            .resctrl
+            .reconcile_group(&group_path, || Ok(pids.clone()), passes);
+
+        let container_ok = matches!(res, Ok(ar) if ar.missing == 0);
+
+        // Update state and emit event with counts
+        // Update state and emit event with counts
+        let (total, reconciled) = {
+            let mut st = self.state.lock().unwrap();
+            // Ensure pod entry and group path
+            st.pods
+                .entry(pod_uid.clone())
+                .or_default()
+                .group_path = Some(group_path.clone());
+
+            // Prior reconciled status
+            let _was_reconciled = st
+                .containers
+                .get(&container.id)
+                .map(|c| c.reconciled)
+                .unwrap_or(false);
+
+            // Update/insert container state
+            st.containers.insert(
+                container.id.clone(),
+                ContainerState { pod_uid: pod_uid.clone(), reconciled: container_ok },
+            );
+
+            // Recompute counts
+            let total = st
```

Comment:
no need to recompute -- set the current container's reconcile state based on the return from reconcile_group and then increment the relevant counters on the pod